README
cvs2ticker README
=================

Installation
------------
Please read the INSTALL file accompanying this file.

What is cvs2ticker?
-------------------
cvs2ticker is an Elvin application designed to improve awareness between
teams of programmers or other people working together on a project stored
in a CVS repository.

NOTE: cvs2ticker requires both Elvin and CVS; for more information see below.

cvs2ticker is designed to be hooked into CVS such that it is run after each
modification to the repository, typically during `cvs commit' operations.
cvs2ticker extracts relevant information (e.g. module, modified files, and the
user-supplied comment) and sends out an Elvin notification containing this
information. The Elvin notification includes the attributes used by the
Tickertape application (see below), allowing Tickertape users to be notified
of the CVS events.

cvs2ticker is accompanied by a CGI script, cvs2web, that displays the details
of the CVS event. cvs2ticker encodes the relevant information and sends it
as arguments in an attached URL with the Elvin notification. The URL points to
the cvs2web CGI script, which decodes the information in order to display it.
cvs2web allows simple operations including viewing the latest version of the
file (obtained from the repository), viewing the differences between the latest
version and the previous version, and viewing the CVS log information. It also
provides a link to a CVSWeb (see below), providing other information and
operations. The net effect is that Tickertape users can simply select a CVS
notification and immediately view all the relevant information in their
browsers.

What is Elvin?
--------------
Elvin is a content-addressed message routing system.  It is often used
for notification of events in distributed applications, and works well
as a communications middleware.

More information about the architecture, software and applications is
available at

    http://www.elvin.org/

What is Tickertape?
-------------------
Tickertape is an Elvin application that uses a scrolling one-line window
to provide users with a wide variety of information from many sources using a
minimum of screen space.

More information about Tickertape, and implementations on several platforms
are available from the Tickertape web site at

    http://www.tickertape.org/

What is CVS?
------------
CVS is the Concurrent Versions System, a version control system in widespread
use for software development. Although useful for individuals, the key
benefit is allowing many people in potentially widely distributed locations to
work on the same set of files concurrently.

More information about CVS, and implementations on most major platforms
are available from the CVS web page at

    http://www.cvshome.org/

What is CVSWeb?
---------------
CVSWeb is an intuitive web interface to CVS. It allows easy browsing of a CVS
repository and access to current and old versions of the files in it. It has
many nice features including colourful, side-by-side diffs between arbitrary
versions of a file.

More information about CVSWeb, as well as CVSWeb itself, is available from
the CVSWeb web page at

    http://stud.fh-heilbronn.de/~zeller/cgi/cvsweb.cgi/


@(#) $Id: README,v 1.5 2006/07/04 09:30:29 d Exp $


NEWS
NEWS

04-Jul-2006
- 1.5.0
  - First release from tickertape.org

17-Jan-2006
- 1.4.1
  - Distribution and installation is now much improved
  - Tickertape v3 message attributes added to emitted
    notifications
  - Released to tickertape.org under a BSD license

27-Nov-2002	Ian Lister <ilister@dstc.com>
- 1.4.0
  - Directory names are included in the notification
  - cvs2web URLs are now encoded more safely

06-Aug-2001	Bill Segall <bill@segall.net>
- 1.3.0
  - Use distutils for packaging
  - python-elvin-4.1b1 compatibility


cvs2ticker.py
#! /usr/bin/env python
########################################################################
# COPYRIGHT_BEGIN
#
#              Tickertape
#              cvs loginfo producer
#
# File:        $Source: /home/d/work/personal/ticker-cvs/cvs2ticker/cvs2ticker.py,v $
# Version:     $Id: cvs2ticker.py,v 1.39 2006/07/04 09:30:29 d Exp $
#
# Copyright    (C) 1998-2006 Mantara Software
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
# * Redistributions of source code must retain the above
#   copyright notice, this list of conditions and the following
#   disclaimer.
#
# * Redistributions in binary form must reproduce the above
#   copyright notice, this list of conditions and the following
#   disclaimer in the documentation and/or other materials
#   provided with the distribution.
#
# * Neither the name of Mantara Software nor the names
#   of its contributors may be used to endorse or promote
#   products derived from this software without specific prior
#   written permission. 
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# REGENTS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
# COPYRIGHT_END
########################################################################
"""

cvs2ticker - pass CVS loginfo messages through to tickertape

"""
__author__ = 'ticker-user@tickertape.org'
__version__ = "$Revision: 1.39 $"[11:-2]


########################################################################
########################################################################
#
#  CONFIGURATION SECTION
#

DEFAULT_GROUP = "CVS"
TIMEOUT       = 10
CVS2WEB_URL   = "http://www.tickertape.org/cgi-bin/cvs2web.py"


#  end of configuration
########################################################################
########################################################################

import base64, os, pickle, sys, getopt, random, string, time, urllib
import elvin


########################################################################

LOG_MESSAGE    = 'Log Message:'
MODIFIED_FILES = 'Modified Files:'
ADDED_FILES    = 'Added Files:'
REMOVED_FILES  = 'Removed Files:'
TEXT_UPDATE    = 'Update of '
TEXT_INDIR     = 'In directory '

VENDOR_TAG     = 'Vendor Tag:'
VENDOR_KEY     = 'Vendor-Tag'
RELEASE_TAG    = 'Release Tags:'
RELEASE_KEY    = 'Release-Tags'
PLAIN_TAG      = '      Tag:'
PLAIN_KEY      = 'Tag'

TEXT_IMPORT    = 'cvs import:'
STATUS         = 'Status:'
IMPORTED_KEY   = 'Imported-Files'

d_section     = {LOG_MESSAGE:    "Log-Message",
                 MODIFIED_FILES: "Modified-Files", 
                 ADDED_FILES:    "Added-Files",
                 REMOVED_FILES:  "Removed-Files",
                 STATUS:         "Status",
                 }

USAGE = "Usage: %s [options] path\n" \
        "[-e elvin-url]  Elvin server if you want to be spcific\n" \
        "[-g group]      Tickertape group for notifications\n" \
        "[-r group]      Tickertape group responses should go to\n" \
        "[-n name]       friendly name for the repository\n" \
        "path            absolute path to repository files\n"


########################################################################

def GetUserName():
    """Find and return the user name

    Raises Exc_noname"""

    if os.environ.has_key('LOGNAME'):
        user = os.environ['LOGNAME']
    elif os.environ.has_key('USER'):
        user = os.environ['USER']
    else:
        raise Exc_noname, "Can't get user name"

    return user


def log_to_ticker(ticker_group, reply_group, repository, rep_dir, bastard):
    """Generate a notification dictionary describing the CVS event.

    *ticker_group*  -- Tickertape group to notify
    *reply_group*   -- Tickertape group replies should go to
    *repository*    -- string repository name
    *rep_dir*       -- absolute path to repository
    Returns         -- dictionary for Elvin notification

    This is pretty ugly ... one day i should tidy up dl's old stuff.
    """

    #-- initialise parsed info
    d_notify = {}
    cur_section = None
    extratext = ""
    found_files = 0    # whether there have been any added/removed/etc

    for key in d_section.values():
        d_notify[key] = ""

    #-- read and process log message
    lines = sys.stdin.readlines()
    for line in lines:

        #-- remove trailing newline
        if line[-1] == "\n":
            line = line[:-1]

        #-- skip blank lines
        if not string.strip(line):
            continue

        if line in d_section.keys():
            #-- handle multi-line sections
            cur_section = d_section[line]

        elif line[0] == "\t":
            found_files = 1
            d_notify[cur_section] = d_notify[cur_section] + ' ' + string.strip(line)

        elif string.find(line, VENDOR_TAG) == 0:
            cur_section = None
            d_notify[VENDOR_KEY] = line[len(VENDOR_TAG)+1:]

        elif string.find(line, RELEASE_TAG) == 0:
            cur_section = None
            if not d_notify.has_key(RELEASE_KEY):
                d_notify[RELEASE_KEY] = string.strip(line[len(RELEASE_TAG)+1:])
            else:
                d_notify[RELEASE_KEY] = d_notify[RELEASE_KEY] + ' ' + string.strip(line[len(PLAIN_TAG)+1:])
        elif string.find(line, PLAIN_TAG) == 0:
            if not d_notify.has_key(PLAIN_KEY):
                d_notify[PLAIN_KEY] = string.strip(line[len(PLAIN_TAG)+1:])
            else:
                d_notify[PLAIN_KEY] = d_notify[PLAIN_KEY] + ', ' + string.strip(line[len(PLAIN_TAG)+1:])
        elif cur_section == d_section[LOG_MESSAGE]:
            d_notify[cur_section] = d_notify[cur_section] + ' ' + line

        else:
            cur_section = None

            #-- process remaining lines
            if string.find(line, TEXT_INDIR) == 0:
                d_notify["Working-Directory"] = line[len(TEXT_INDIR):]

            elif string.find(line, TEXT_UPDATE) == 0:
                d_notify["Repository-Directory"] = os.path.normpath(line[len(TEXT_UPDATE):])

            elif line[0:2] == "N ":
                if not d_notify.has_key(IMPORTED_KEY):
                    d_notify[IMPORTED_KEY] = line[2:]
                else:
                    d_notify[IMPORTED_KEY] = d_notify[IMPORTED_KEY] + ' ' + line[2:]

            elif line[0:2] == "I ":
                #-- ignore these, since CVS does ...
                pass

            else:
                extratext = extratext + ' ' + line

    #-- add non-parsed text
    d_notify["Extras"] = extratext
    d_notify["Original"] = reduce(lambda s,e: s+e, lines, "")
    d_notify["Repository"] = repository
    d_notify["Repository-Root"] = rep_dir
    str_dir = d_notify["Repository-Directory"]
    rep_rel_path = str_dir[len(rep_dir)+1:]
    module = string.replace(string.split(rep_rel_path, "/")[0], '+', '%2b')
    d_notify["Relative-Directory"] = rep_rel_path
    d_notify["Module"] = module

    #-- create tickertape message
    msg = "In %s:" % rep_rel_path

    if d_notify[d_section[ADDED_FILES]]:
        msg = msg + " Added" + d_notify[d_section[ADDED_FILES]]

    if d_notify[d_section[REMOVED_FILES]]:
        msg = msg + " Removed" + d_notify[d_section[REMOVED_FILES]]

    if d_notify[d_section[MODIFIED_FILES]]:
        msg = msg + " Modified" + d_notify[d_section[MODIFIED_FILES]]

    if d_notify.has_key(IMPORTED_KEY):
        msg = msg + " Import"

    if d_notify.has_key(PLAIN_KEY):
        msg = msg + " (tag " + d_notify[PLAIN_KEY] + ")"

    #-- the bill trap
    if bastard:
        if not string.strip(d_notify[d_section[LOG_MESSAGE]]):
            d_notify[d_section[LOG_MESSAGE]] = "%s, the slack bastard, didn't supply a log message." % user

    if found_files:
        msg = msg + ':' + d_notify[d_section[LOG_MESSAGE]]
    else:
        msg = msg + d_notify[d_section[LOG_MESSAGE]]

    #-- create attachment URL
    str_url = CVS2WEB_URL
    str_url = str_url + "?%s+%s" % (user, urllib.quote(pickle.dumps(d_notify)))

    #-- add tickertape-specific attributes
    d_notify.update({'TIMEOUT' : TIMEOUT,
                     'TICKERTEXT' : msg,
                     'TICKERTAPE' : ticker_group,
                     'USER' : user,
                     'MIME_TYPE':   "x-elvin/url",
                     'MIME_ARGS':   str_url,
                     'Message-Id':  str(random.randint(1, 0x7ffffff))})

    #-- add v3 tickertape attributes
    d_notify.update({'Timeout': d_notify['TIMEOUT'] * 60,
                     'Message': d_notify['TICKERTEXT'],
                     'Group': d_notify['TICKERTAPE'],
                     'From': d_notify['USER']})

    if reply_group:
        d_notify['Reply-To'] = reply_group;

    d_notify['Attachment'] = 'MIME-Version: 1.0\r\n' \
                             'Content-Type: text/uri-list\r\n' \
                             '\r\n' \
                             '%s\r\n' % str_url
    return d_notify


def error_exit(msg):
    """Print error message and exit."""

    #-- get executable name
    progname = os.path.basename(sys.argv[0])

    #-- message
    sys.stderr.write(USAGE % progname)
    sys.stderr.write("\n%s\n\n" % msg)

    #-- quit
    sys.exit(1)


########################################################################

if __name__ == '__main__':

    #-- initialise
    urls = []
    group = None
    reply_group = None
    repository = None
    d_notify = None
    bastard = 1

    #-- check mandatory args
    if len(sys.argv) < 3:
        error_exit("Not enough arguments.")

    rep_dir = sys.argv[-1]

    #-- parse options
    try:
        (optlist,args) = getopt.getopt(sys.argv[1:-1], "e:g:n:r:b")
    except:
        error_exit("Failed to process the arglist: %s" % str(sys.argv[1:-1]))

    for (opt, arg) in optlist:
        if opt == '-e':
            urls.append(arg)

        if opt == '-g':
            if not group:
                group = arg
            else:
                error_exit("Only one group specification allowed.")

        if opt == '-n':
            if not repository:
                repository = arg
            else:
                error_exit("Only one name specification allowed.")

        if opt == '-r':
            if not reply_group:
                reply_group = arg
            else:
                error_exit("Only one reply group specification allowed.")

        if opt == '-b':
            bastard = 0

    #-- set default option values
    if not group:
        group = DEFAULT_GROUP

    if not repository:
        repository = rep_dir

    c = elvin.client()
    e = c.connection()

    for url in urls:
        e.append_url(url)
        e.set_discovery(0)
    else:
        e.set_discovery(1)

    e.open()

    #-- get user
    user = GetUserName()

    #-- parse log message
    d_notify = log_to_ticker(group, reply_group, repository, rep_dir, bastard)
    if d_notify:
        e.notify(d_notify)

    sys.exit(0)


########################################################################


cvs2web.py
#! /usr/bin/env python
########################################################################
# COPYRIGHT_BEGIN
#
#              Tickertape
#              cvs loginfo producer
#
# File:        $Source: /home/d/work/personal/ticker-cvs/cvs2ticker/cvs2web.py,v $
# Version:     $Id: cvs2web.py,v 1.22 2006/07/04 09:30:29 d Exp $
#
# Copyright    (C) 1999-2006 Mantara Software
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
# * Redistributions of source code must retain the above
#   copyright notice, this list of conditions and the following
#   disclaimer.
#
# * Redistributions in binary form must reproduce the above
#   copyright notice, this list of conditions and the following
#   disclaimer in the documentation and/or other materials
#   provided with the distribution.
#
# * Neither the name of Mantara Software nor the names
#   of its contributors may be used to endorse or promote
#   products derived from this software without specific prior
#   written permission. 
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# REGENTS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
# COPYRIGHT_END
########################################################################
"""

cvs2web - takes CVS commit information encoded by cvs2ticker and presents
it on the web. Allows viewing the file, diff, log or linking to cvsweb.


"""
__author__  = "ticker-user@tickertape.org"
__version__ = "$Revision: 1.22 $"[11:-2]

#############################################################################
#############################################################################
#
#  CONFIGURATION SECTION
#

CGI_URL       = "/cgi-bin"
CVS2WEB_URL   = CGI_URL + "/cvs2web.py"

# Set BUGZILLA_URL to the empty string if you don't use Bugzilla
BUGZILLA_URL = "http://www.tickertape.org/bugzilla/show_bug.cgi?id=%s"
# BUGZILLA_URL = ""

# Set CVSWEB_URL to the empty string if you don't use CVSWeb or ViewCVS
CVSWEB_URL    = CGI_URL + "/cvsweb.cgi"
# CVSWEB_URL    = ""

# Set TICKER_URL to the empty string if you don't use CGITicker
TICKER_URL    = CGI_URL + "/ticker.py"
# TICKER_URL    = ""

# Set MAILDOMAIN to the empty string if all users don't share a domain
# for mailto links
MAIL_DOMAIN   = "tickertape.org"
# MAIL_DOMAIN   = ""

# Set RCSPATH to the location of the installed RCS tools (with a
# trailing slash), or empty to use the default PATH
RCSPATH = ""
# RCSPATH = "/usr/local/bin/"

#  end of configuration
#############################################################################
#############################################################################

import base64, cgi, exceptions, os, pickle, popen2, regsub, string, sys, time, traceback, urllib


#############################################################################

def send(txt, indent=0):
    """Write text to output."""
    
    #-- create indent string
    str_indent = " " * indent

    #-- insert indent after carriage returns
    txt = regsub.gsub("\n\(.\)", "\n%s\\1" % str_indent, txt)

    #-- write
    sys.stdout.write("%s%s" % (str_indent, txt))
    
    return


def wrap(s, width=60):
    """Return the string with long strings wrapped."""

    res = ""
    
    #-- separate into paragraphs
    lst_para = string.split(s, "\n")
    for para in lst_para:
        if len(para) <= width:
            res = res + para + "\n"

        else:
            lst_word = string.split(para)
            line = ""
            for word in lst_word:
                if not line:
                    line = word
                    
                elif len(line) + len(word) + 1 <= width:
                    line = line + " " + word
                    
                else:
                    res = res + line + "\n"
                    line = word
                    
            res = res + line + "\n"

    return res


def common_tail(str1, str2):
    """ """
    return ""

            
#############################################################################

def header():
    send("Content-type: text/html\n\n")
    send("<html>\n")
    send("<head>\n", 2)
    send("<title>cvs2ticker</title>\n", 4)
    send("</head>\n", 2)

    send('<body bgcolor="white">\n', 2)
    send('<h1><tt>cvs2web, v%s</tt><hr></h1>\n' % __version__, 4)
    
    return


def end_body():
    send("<hr>\n", 4)
    str_time = time.asctime(time.localtime(time.time()))
    
    send("<address>%s</address>\n" % str_time, 4)
    send("</body>\n", 2)
    send("</html>\n\n")
    return


#############################################################################

def user_info(user, d_cvs):
    """Show the name of the user who performed the commit"""

    #-- get the name of the person?
    logname = "cvs%20user"

    #-- find the module name(s)
    str_dir = d_cvs["Repository-Directory"]
    str_rep = d_cvs["Repository-Root"]
    str_rep_name = d_cvs["Repository"]
    rep_rel_path = str_dir[len(str_rep)+1:]
    mod_rel_path = string.join(string.split(rep_rel_path, "/")[1:], "/")
    module = string.split(rep_rel_path, "/")[0]

    send('<dl>\n', 4)
    send('<dt><b>cvs commit</b></tt><p>\n', 6)
    send('<dd>\n  <table>\n', 6)
    send('<tr><td>Repository</td> <td><b>%s</b></td> <td>&nbsp;</td></tr>\n' % str_rep_name,10)
    send('<tr><td>Module</td> <td><b>%s</b></td> <td>&nbsp;</td></tr>\n' % module, 10)
    send('<tr><td>User</td> <td><b>%s</b></td>\n' % user, 10)
    send('<td>\n', 14)
    if MAIL_DOMAIN:
        send('[<a href="mailto:%s@%s">mail</a>]\n' % (user, MAIL_DOMAIN), 18)
    if TICKER_URL:
        send('[<a href="%s?%s+10+%s+%s+%s">ticker</a>]\n' % \
         (TICKER_URL, logname, user, module, str_rep_name), 18)
    send('</td>\n', 14)
    send('</tr>\n', 10)
    send('</table>\n', 8)
    send('</dl>\n', 4)
    send('<p>\n', 4)
    return


def log_msg(d_cvs):
    if not d_cvs.has_key("Log-Message"):
        return

    send('<table align="center" width="90%">\n', 4)
    send('<tr>\n', 6)
    send('<td bgcolor="#c0c0c0">\n', 8)
    send("<pre>\n\n", 10)

    if BUGZILLA_URL:
        bugzilla_url = BUGZILLA_URL % ('\\2')
        replacement = '<a href="%s">\\1</a>' % (bugzilla_url)
        send(regsub.gsub('\([Bb]ug #?\([0-9]+\)\)', replacement, wrap(d_cvs["Log-Message"])))  
        
    else:
	send(wrap(d_cvs["Log-Message"]))

    send("\n</pre>\n", 10)
    send("    </td>\n  </tr>\n</table>\n", 4)
    return


def add_info(d_cvs):

    if not d_cvs.has_key("Added-Files"):
        return

    if not d_cvs["Added-Files"]:
        return
    
    #-- find the module name(s)
    str_dir = d_cvs["Repository-Directory"]
    str_rep = d_cvs["Repository-Root"]
    str_rep_name = d_cvs["Repository"]
    rep_rel_path = str_dir[len(str_rep)+1:]
    mod_rel_path = string.join(string.split(rep_rel_path, "/")[1:], "/")
    
    send("<dl>\n  <dt>Added files:\n", 4)

    for file in string.split(d_cvs["Added-Files"]):
        full_path = os.path.join(str_dir, file)
        
        send("<dd>%s/%s\n" % (mod_rel_path, file), 6)
        send('[<a href="%s?file+%s">file</a>]\n' % (CVS2WEB_URL, full_path), 10)
        if CVSWEB_URL:
            send('[<a href="%s/%s/%s?cvsroot=%s">cvsweb</a>]\n' % (CVSWEB_URL, rep_rel_path, file, str_rep_name), 10)
        
    send("</dl>", 4)
    send("<p>", 4)
    
    return


def modify_info(d_cvs):
    """Print info on modified files."""

    if not d_cvs.has_key("Modified-Files"):
        return

    if not d_cvs["Modified-Files"]:
        return
    
    send("<dl>\n  <dt>Modified files:\n", 4)

    #-- find the module name(s)
    str_dir = d_cvs["Repository-Directory"]
    str_rep = d_cvs["Repository-Root"]
    str_rep_name = d_cvs["Repository"]
    rep_rel_path = str_dir[len(str_rep)+1:]
    mod_rel_path = string.join(string.split(rep_rel_path, "/")[1:], "/")
    
    for file in string.split(d_cvs["Modified-Files"]):
        full_path = os.path.join(str_dir, file)
        
        send('<dd>%s/%s\n' % (mod_rel_path, file), 6)
        send('[<a href="%s?file+%s">file</a>]\n' % (CVS2WEB_URL, full_path), 10)
        send('[<a href="%s?diff+%s">diff</a>]\n' % (CVS2WEB_URL, full_path), 10)
        send('[<a href="%s?log+%s">log</a>]\n' % (CVS2WEB_URL, full_path), 10)
        if CVSWEB_URL:
            send('[<a href="%s/%s/%s?cvsroot=%s">cvsweb</a>]\n' % (CVSWEB_URL, rep_rel_path, file, str_rep_name), 10)
        
    send("</dl>\n", 4)
    send("<p>\n", 4)
    
    return


def import_info(d_cvs):
    if not d_cvs.has_key("Imported-Files"):
        return

    if not d_cvs["Imported-Files"]:
        return

    #-- find the repository details
    str_rep = d_cvs["Repository-Root"]
    str_rep_name = d_cvs["Repository"]
    
    send("<dl>\n  <dt>Imported files:\n", 4)

    for file in string.split(d_cvs["Imported-Files"]):
        full_path = os.path.join(str_rep, file)

        send('<dd>%s\n' % file, 6)
        send('[<a href="%s?file+%s">file</a>]\n' % (CVS2WEB_URL, full_path), 10)
        send('[<a href="%s?log+%s">log</a>]\n' % (CVS2WEB_URL, full_path), 10)
        if CVSWEB_URL:
            send('[<a href="%s/%s?cvsroot=%s">cvsweb</a>]\n' % (CVSWEB_URL, file, str_rep_name), 10)

    send("</dl>\n", 4)
    send("<p>\n", 4)
    
    return


def remove_info(d_cvs):
    if not d_cvs.has_key("Removed-Files"):
        return

    if not d_cvs["Removed-Files"]:
        return
    
    send("<dl>\n  <dt>Removed files:\n", 4)

    for file in string.split(d_cvs["Removed-Files"]):
        send("<dd>%s\n" % file, 6)
        
    send("</dl>", 4)
    send("<p>", 4)
    
    return


#############################################################################

def file(str_file):
    """Return the full source of the named file."""

    send('Content-type: text/plain\n\n')

    try:
        os.system("cd /tmp; " + RCSPATH + "co %s,v" % str_file)
    
        f = open("/tmp/%s" % os.path.basename(str_file))
        s = f.read()
        f.close()

        if s:
            send(s)

        else:
            send("< zero length file >")
            
    finally:
        os.unlink("/tmp/%s" % os.path.basename(str_file))
        
    return


#############################################################################

def diff(str_file):
    """Return the diff between the current and last versions."""

    send('Content-type: text/plain\n\n')

    try:
        #-- check out working file in /tmp
        os.system("cd /tmp; " + RCSPATH + "co %s,v" % str_file)

        #-- run rlog to get latest version number
        p = os.popen(RCSPATH + "rlog %s" % str_file, "r")
        l = p.readlines()
        p.close()

        rev = l[3][6:-1]
        major, minor = string.split(rev, ".")
        prev = "%s.%s" % (major, str(string.atoi(minor) - 1))

        #-- run rcsdiff
        pout, pin, perr = popen2.popen3("cd /tmp; " + RCSPATH + "rcsdiff -r%s -u %s,v 2>&1" % (prev, str_file))
        s = pout.read()
        pout.close()
        pin.close()
        perr.close()
        
        #-- return output
        send(s)

    finally:
        #-- remove checked out working file
        os.unlink("/tmp/%s" % os.path.basename(str_file))

    return


#############################################################################

def log(str_file):
    """Return the log for the file."""

    send('Content-type: text/plain\n\n')
    try:
        p = os.popen(RCSPATH + "rlog %s" % str_file, "r")
        s = p.read()
        p.close()

        send(s)

    finally:
        sys.exit(0)
        
    return


#############################################################################

def error(msg, e=None):
    """Print a HTML usage message."""
    
    send('Content-type: text/plain\n\n')
    send('cvs2web, v%s\n' % __version__)
    send('%s\n\n' % msg)

    send('Required arguments are ...\n')
    send('user  -- string user name\n')
    send('dict  -- pickled dictionary of parsed CVS log message\n\n')

    send('Traceback\n')
    if e:
        for s in traceback.format_exception(sys.exc_type, sys.exc_value, sys.exc_traceback):
            send(s)

    return

        
#############################################################################

if __name__ == "__main__":

    #-- check parameters
    if len(sys.argv) != 3:
        error("Not enough arguments: %d, %s" % (len(sys.argv), str(sys.argv)))
        sys.exit(0)

    #-- process arguments
    user = sys.argv[1]

    if user == "log":
        log(sys.argv[2])
        sys.exit(0)

    elif user == "diff":
        diff(sys.argv[2])
        sys.exit(0)

    elif user == "file":
        file(sys.argv[2])
        sys.exit(0)

    elif user == "commit":
        pass

    else:
        try:
            raw_str = sys.argv[2]

            if raw_str[0] == "\\" or raw_str[0] == "%":
                # new-style escaping
                raw_str = urllib.unquote(raw_str)
                str_pkl = string.replace(raw_str, "\\", "")

            else:
                # backward compatibility.  \x escapes are no-longer generated
                raw_str = string.replace(raw_str, r'\x0a', '\012')
                raw_str = string.replace(raw_str, r'\x3d', '=')
                raw_str = string.replace(raw_str, r'\x2b', '+')

                str_pkl = base64.decodestring(raw_str)

            # unpack
            d_cvs = pickle.loads(str_pkl)

        except exceptions.Exception, e:
            error("URL decoding error", e)
            send('Pickle arg\n')
            send(sys.argv[2])
            sys.exit(0)

        #-- start HTML output
        header()

        user_info(user, d_cvs)

        log_msg(d_cvs)

        import_info(d_cvs)
        add_info(d_cvs)
        remove_info(d_cvs)
        modify_info(d_cvs)
        
        end_body()
        sys.exit(0)

    
#############################################################################

